File: dsl-api/src/main/java/cbs/dsl/api/WorkflowTypes.java
package cbs.dsl.api;

import cbs.dsl.api.WorkflowFunction.WorkflowArg;
import cbs.dsl.api.WorkflowFunction.WorkflowResult;
import io.avaje.jsonb.Json;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.NoArgsConstructor;

import java.util.List;
import java.util.Map;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class WorkflowTypes {

  @Json
  @Builder(toBuilder = true)
  public record WorkflowInput(
      String currentState, String action, Map<String, Object> params, String workflowInstanceId)
      implements WorkflowArg {

  }

  @Json
  @Builder(toBuilder = true)
  public record WorkflowOutput(String nextState, List<String> events, String status)
      implements WorkflowResult {

    public WorkflowOutput(String nextState) {
      this(nextState, List.of(), "SUCCESS");
    }

  }
}

File: dsl-api/src/main/kotlin/cbs/dsl/api/WorkflowTypes.kt
package cbs.dsl.api

import io.avaje.jsonb.Json

/**
 * Input carrier for workflow state transition execution. Carries the current state, action to
 * perform, parameters, and optional workflow instance metadata as JSON-serializable data for
 * Temporal workflow replay and PostgreSQL JSONB storage.
 */
@Json
data class WorkflowInput(
    val currentState: String,
    val action: String,
    val params: Map<String, Any?> = emptyMap(),
    val workflowInstanceId: String? = null,
)

/**
 * Output carrier for workflow state transition execution. Carries the next state, event codes to
 * execute, and status as JSON-serializable data for Temporal workflow replay and PostgreSQL JSONB
 * storage.
 */
@Json
data class WorkflowOutput(
    val nextState: String,
    val events: List<String> = emptyList(),
    val status: String = "SUCCESS",
)
